src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -77,16 +111,34 @@ internal class AnalyzerManager
                 }, cancellationToken);
             };
 
-            var task = ImmutableInterlocked.GetOrAdd(ref _sessionScopeMap, analyzer, getTask);
+            var callback = new ConditionalWeakTable<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.CreateValueCallback(getTask);
+            return _sessionScopeMap.GetValue(analyzer, callback);
+        }
 
-            // Retry cancelled task.
-            if (task.Status == TaskStatus.Canceled)
+        private async Task<HostSessionStartAnalysisScope> GetSessionAnalysisScopeAsync(
+            DiagnosticAnalyzer analyzer,
+            Action<Diagnostic> addDiagnostic,
+            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
+            CancellationToken cancellationToken)
+        {
+            try
             {
-                ImmutableInterlocked.TryUpdate(ref _sessionScopeMap, analyzer, getTask(analyzer), task);
-                return _sessionScopeMap[analyzer];
+                return await GetSessionAnalysisScopeCoreAsync(analyzer, addDiagnostic, continueOnAnalyzerException, cancellationToken).ConfigureAwait(false);
             }
+            catch (TaskCanceledException)
+            {
+                // Task to compute the scope was cancelled.
+                // Clear the entry in scope map for analyzer, so we can attempt a retry.
+                _sessionScopeMap.Remove(analyzer);
 
-            return task;
+                // Don't attempt retry if our cancellation token fired as the retry will also fail.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return new HostSessionStartAnalysisScope();
```

Comment:
As above, why is it necessary to create an object if cancellation has actually been requested?